app/src/main/java/com/github/epfl/meili/poi/PoiGoogleRetriever.kt

```diff
@@ -0,0 +1,112 @@
+            queryGooglePlacesAPI(query, onSuccess, onError)
+        }
+    }
+
+    fun queryGooglePlacesAPI(query: String, onSuccess: (List<PointOfInterest>) -> Unit, onError: (VolleyError) -> Unit) {
+
+        val jsonObjectRequest = JsonObjectRequest(Request.Method.GET, GOOGLE_PLACES_URL + query,
+                null, customOnSuccessFrom(onSuccess), onError)
+
+        queue.add(jsonObjectRequest)
+    }
+
+    fun customOnSuccessFrom(onSuccess: (List<PointOfInterest>) -> Unit): (JSONObject) -> Unit {
+        return { response ->
+            if (response["status"] == "OK") {
+                val placesResponse = Gson().fromJson(response.toString(), GooglePlacesResponse::class.java)
+                Log.d(TAG, response.toString())
+                Log.d(TAG, placesResponse.getCustomPois().toString())
+                onSuccess(placesResponse.getCustomPois())
+            }
+        }
+    }
+
+    companion object {
+        private const val GOOGLE_PLACES_URL = "https://maps.googleapis.com/maps/api/place/nearbysearch/json?"
+        private const val MAX_COVERAGE_RADIUS = 3000
+        private const val TAG = "PoiGoogleRetriever"
+    }
+}
+
+data class GooglePlacesResponse(
+        @SerializedName("results")
+        val pointsOfInterest: List<PlacesPointOfInterest> = ArrayList()
+) {
+    fun getCustomPois(): List<PointOfInterest> {
+        val poiList = ArrayList<PointOfInterest>()
+        for (poi in pointsOfInterest) {
+            if (poi.geometry != null && poi.name != null && poi.geometry.latLng != null && poi.uid != null) {
+                poiList.add(poi.toStandardPoi())
+            }
+        }
+        return poiList
+    }
+}
+
+//TODO: we could also get some photos from the GooglePlaces API if we want
+data class PlacesPointOfInterest(
+        @SerializedName("geometry")
+        val geometry: PoiGeometry? = null,
+        @SerializedName("place_id")
+        val uid: String? = null,
+        @SerializedName("name")
+        val name: String? = null,
+        @SerializedName("icon")
+        val icon: String? = null,
+        @SerializedName("types")
+        val poiTypes: List<String>? = null,
+        @SerializedName("opening_hours")
+        val openingHours: PoiOpeningHours? = null
+) {
+    fun toStandardPoi(): PointOfInterest {
+        if (openingHours != null) {
+            return PointOfInterest(geometry!!.latLng!!.latitude!!, geometry.latLng!!.longitude!!, name!!, uid!!, icon!!, poiTypes!!, openingHours.openNow)
+        }
+        return PointOfInterest(geometry!!.latLng!!.latitude!!, geometry.latLng!!.longitude!!, name!!, uid!!, icon!!, poiTypes!!)
+    }
+}
+
+data class PoiGeometry(
+        @SerializedName("location")
+        val latLng: PoiLocation? = null
+)
+
+data class PoiLocation(
+        @SerializedName("lat")
+        val latitude: Double? = null,
+        @SerializedName("lng")
+        val longitude: Double? = null
+)
+
```

Comment:
Can't we use LatLng (https://developers.google.com/android/reference/com/google/android/gms/maps/model/LatLng) for this?

Reply:
Thanks for your comments. The thing is that this class in only used for the json deserialisation, so if I put LatLng it won't deserialise the object correctly and instead I have to get each value separately (latitude and longitude).